addons/delivery/models/stock_picking.py

```diff
@@ -187,6 +187,17 @@ def _add_delivery_cost_to_so(self):
         sale_order = self.sale_id
         if sale_order.invoice_shipping_on_delivery:
             sale_order._create_delivery_line(self.carrier_id, self.carrier_price)
+        else:
+            delivery_line = sale_order.order_line.filtered(lambda line: line.is_delivery and line.price_unit == 0)
+            if delivery_line:
+                res = self.carrier_id.rate_shipment(sale_order)
+                if res['success']:
+                    delivery_line.qty_delivered = 1  # to trigger the invoice status
+                    delivery_line.price_unit = res['price']
+                    # remove the estimated price from the description
+                    delivery_line.name = delivery_line.name.split('(Estimated')[0]
```

Comment:
won't work with translatin
you could re-write carrier.with_context(lang=self.partner_id.lang).name i guess